File: test/services/shipstation/create_order_test.rb
require "test_helper"

# Stand-in for the Fluid orders client so #call doesn't hit the network when it
# writes the external id back to Fluid.
class FakeOrdersService
  def update_external_id(*, **)
    true
  end
end

# Simulates a Fluid external-id sync failure (must not fail the order).
class RaisingOrdersService
  def update_external_id(*, **)
    raise "fluid timeout"
  end
end

# Stand-in for Shipstation::CancelOrder returning a fixed outcome.
class FakeCancel
  def initialize(result)
    @result = result
  end

  def call(_shipstation_order_id)
    @result
  end
end

class Shipstation::CreateOrderTest < ActiveSupport::TestCase
  fixtures :companies

  setup do
    @company = companies(:acme)
    @company.create_integration_setting!(settings: { "api_key" => "k", "api_secret" => "s" })
  end

  def order_params(metadata: {}, status: nil, id: 555, order_number: "ORD-555")
    order = {
      "id" => id,
      "order_number" => order_number,
      "email" => "a@b.com",
      "items" => [],
      "ship_to" => { "name" => "X" },
      "metadata" => metadata,
    }
    order["status"] = status if status
    { "company_id" => @company.fluid_company_id, "order" => order }
  end

  def run_with_captured_body(params)
    captured = nil
    post = lambda do |*args, **_kw|
      captured = JSON.parse((args[1] || {})[:body])
      FakeResponse.new({ "orderId" => 42 })
    end
    HTTParty.stub(:post, post) do
      FluidApi::V2::OrdersService.stub(:new, FakeOrdersService.new) do
        Shipstation::CreateOrder.new(params).call
      end
    end
    captured
  end

  # Runs #call and reports whether an HTTP POST to ShipStation was attempted.
  def run_tracking_http(params)
    posted = false
    HTTParty.stub(:post, ->(*_a, **_k) { posted = true; FakeResponse.new({ "orderId" => 42 }) }) do
      FluidApi::V2::OrdersService.stub(:new, FakeOrdersService.new) do
        Shipstation::CreateOrder.new(params).call
      end
    end
    posted
  end

  test "injects mapped carrier/service/package codes and the requested service" do
    @company.shipping_method_mappings.create!(
      fluid_shipping_title: "Ground Shipping",
      carrier_code: "stamps_com",
      service_code: "usps_priority_mail",
      package_code: "package",
    )
    body = run_with_captured_body(
      order_params(metadata: { "shipping" => { "title" => "Ground Shipping" } }),
    )
    _(body["carrierCode"]).must_equal "stamps_com"
    _(body["serviceCode"]).must_equal "usps_priority_mail"
    _(body["packageCode"]).must_equal "package"
    _(body["requestedShippingService"]).must_equal "Ground Shipping"
  end

  test "passes the title as requestedShippingService even with no mapping" do
    body = run_with_captured_body(
      order_params(metadata: { "shipping" => { "title" => "Express" } }),
    )
    _(body["requestedShippingService"]).must_equal "Express"
    _(body.key?("carrierCode")).must_equal false
  end

  test "records the seen shipping method with the order number" do
    run_with_captured_body(order_params(metadata: { "shipping" => { "title" => "Express" } }))
    seen = @company.seen_shipping_methods.find_by(fluid_shipping_title: "Express")
    _(seen).wont_be_nil
    _(seen.example_order_number).must_equal "ORD-555"
  end

  test "sends no shipping fields when the order has no shipping title" do
    body = run_with_captured_body(order_params)
    _(body.key?("requestedShippingService")).must_equal false
    _(body.key?("carrierCode")).must_equal false
  end

  # -- status gating / payment hold ----------------------------------------

  test "submits when status is awaiting_shipment" do
    posted = run_tracking_http(order_params(status: "awaiting_shipment"))
    _(posted).must_equal true
    _(@company.shipstation_orders.find_by(fluid_order_id: 555).status).must_equal "SUBMITTED"
  end

  test "submits when status is blank" do
    _(run_tracking_http(order_params)).must_equal true
  end

  test "holds an awaiting_payment order without sending" do
    posted = run_tracking_http(order_params(status: "awaiting_payment"))
    _(posted).must_equal false
    _(@company.shipstation_orders.find_by(fluid_order_id: 555).status).must_equal "AWAITING_PAYMENT"
  end

  test "skips an unfulfillable status without creating a record or sending" do
    posted = run_tracking_http(order_params(status: "cancelled"))
    _(posted).must_equal false
    _(@company.shipstation_orders.find_by(fluid_order_id: 555)).must_be_nil
  end

  test "does not resubmit an order already SUBMITTED" do
    @company.shipstation_orders.create!(fluid_order_id: 555, fluid_order_number: "ORD-555", status: "SUBMITTED")
    posted = run_tracking_http(order_params(status: "awaiting_shipment"))
    _(posted).must_equal false
  end

  test "releases a held order when it becomes fulfillable" do
    # First: held as awaiting_payment.
    run_tracking_http(order_params(status: "awaiting_payment"))
    held = @company.shipstation_orders.find_by(fluid_order_id: 555)
    _(held.status).must_equal "AWAITING_PAYMENT"

    # Then: order.updated arrives fulfillable -> submitted.
    posted = run_tracking_http(order_params(status: "awaiting_shipment"))
    _(posted).must_equal true
    _(held.reload.status).must_equal "SUBMITTED"
  end

  # -- batching hold --------------------------------------------------------

  test "holds a fulfillable order for batching (window sets hold_until)" do
    @company.integration_setting.update!(hold_for_batch: true, batch_window_minutes: 30)
    posted = run_tracking_http(order_params(status: "awaiting_shipment"))
    _(posted).must_equal false
    order = @company.shipstation_orders.find_by(fluid_order_id: 555)
    _(order.status).must_equal "HELD"
    _(order.hold_until).wont_be_nil
  end

  test "holds for batching with no window leaves hold_until nil (manual release)" do
    @company.integration_setting.update!(hold_for_batch: true, batch_window_minutes: nil)
    run_tracking_http(order_params(status: "awaiting_shipment"))
    order = @company.shipstation_orders.find_by(fluid_order_id: 555)
    _(order.status).must_equal "HELD"
    _(order.hold_until).must_be_nil
  end

  test "respect_hold false bypasses batching and submits immediately" do
    @company.integration_setting.update!(hold_for_batch: true, batch_window_minutes: 30)
    posted = false
    HTTParty.stub(:post, ->(*_a, **_k) { posted = true; FakeResponse.new({ "orderId" => 42 }) }) do
      FluidApi::V2::OrdersService.stub(:new, FakeOrdersService.new) do
        Shipstation::CreateOrder.new(order_params(status: "awaiting_shipment"), respect_hold: false).call
      end
    end
    _(posted).must_equal true
    _(@company.shipstation_orders.find_by(fluid_order_id: 555).status).must_equal "SUBMITTED"
  end

  test "awaiting_payment takes precedence over the batching hold" do
    @company.integration_setting.update!(hold_for_batch: true, batch_window_minutes: 30)
    run_tracking_http(order_params(status: "awaiting_payment"))
    _(@company.shipstation_orders.find_by(fluid_order_id: 555).status).must_equal "AWAITING_PAYMENT"
  end

  # -- hardening (review follow-ups) ---------------------------------------

  test "refreshes stored payload on reprocess so a paid held order ships" do
    @company.integration_setting.update!(hold_for_batch: true, batch_window_minutes: 30)
    # Held while awaiting payment.
    run_tracking_http(order_params(status: "awaiting_payment"))
    order = @company.shipstation_orders.find_by(fluid_order_id: 555)
    _(order.status).must_equal "AWAITING_PAYMENT"

    # order.updated arrives paid -> batch-held, payload refreshed.
    run_tracking_http(order_params(status: "awaiting_shipment"))
    order.reload
    _(order.status).must_equal "HELD"
    _(order.request_payload["status"]).must_equal "awaiting_shipment"

    # Release uses the refreshed payload and actually submits.
    posted = false
    HTTParty.stub(:post, ->(*_a, **_k) { posted = true; FakeResponse.new({ "orderId" => 42 }) }) do
      FluidApi::V2::OrdersService.stub(:new, FakeOrdersService.new) do
        payload = { "order" => order.request_payload, "company_id" => @company.fluid_company_id }
        Shipstation::CreateOrder.new(payload, respect_hold: false).call
      end
    end
    _(posted).must_equal true
    _(order.reload.status).must_equal "SUBMITTED"
  end

  test "a Fluid external-id sync failure does not fail or re-mark the order" do
    posted = 0
    HTTParty.stub(:post, ->(*_a, **_k) { posted += 1; FakeResponse.new({ "orderId" => 42 }) }) do
      FluidApi::V2::OrdersService.stub(:new, RaisingOrdersService.new) do
        Shipstation::CreateOrder.new(order_params(status: "awaiting_shipment")).call
      end
    end
    _(posted).must_equal 1
    _(@company.shipstation_orders.find_by(fluid_order_id: 555).status).must_equal "SUBMITTED"
  end

  test "marks an existing pre-submit order CANCELLED when Fluid reports it unfulfillable" do
    @company.integration_setting.update!(hold_for_batch: true, batch_window_minutes: 30)
    run_tracking_http(order_params(status: "awaiting_shipment")) # HELD
    _(@company.shipstation_orders.find_by(fluid_order_id: 555).status).must_equal "HELD"

    run_tracking_http(order_params(status: "cancelled"))
    _(@company.shipstation_orders.find_by(fluid_order_id: 555).status).must_equal "CANCELLED"
  end

  test "preserves the original hold_until across repeated updates" do
    @company.integration_setting.update!(hold_for_batch: true, batch_window_minutes: 30)
    run_tracking_http(order_params(status: "awaiting_shipment"))
    first = @company.shipstation_orders.find_by(fluid_order_id: 555).hold_until
    _(first).wont_be_nil

    run_tracking_http(order_params(status: "awaiting_shipment"))
    second = @company.shipstation_orders.find_by(fluid_order_id: 555).hold_until
    _(second.to_i).must_equal first.to_i
  end

  test "records the seen shipping method even when the order is held for batch" do
    @company.integration_setting.update!(hold_for_batch: true, batch_window_minutes: 30)
    run_tracking_http(order_params(status: "awaiting_shipment", metadata: { "shipping" => { "title" => "Overnight" } }))
    _(@company.shipstation_orders.find_by(fluid_order_id: 555).status).must_equal "HELD"
    _(@company.seen_shipping_methods.find_by(fluid_shipping_title: "Overnight")).wont_be_nil
  end

  # -- store assignment -----------------------------------------------------

  test "assigns the configured store via advancedOptions.storeId" do
    @company.integration_setting.update!(store_id: "3618888")
    body = run_with_captured_body(order_params)
    _(body["advancedOptions"]).must_equal({ "storeId" => 3618888 })
  end

  test "omits advancedOptions when no store is configured" do
    body = run_with_captured_body(order_params)
    _(body.key?("advancedOptions")).must_equal false
  end

  # -- carrier/service pairing + failure handling ---------------------------

  test "drops a carrier with no service and sends requestedShippingService only" do
    # A legacy/incomplete mapping (carrier, no service) that predates the model
    # validation — the service must degrade gracefully rather than push a carrier
    # ShipStation will reject.
    mapping = @company.shipping_method_mappings.build(
      fluid_shipping_title: "Ground Shipping",
      carrier_code: "fedex",
    )
    mapping.save!(validate: false)
    body = run_with_captured_body(
      order_params(metadata: { "shipping" => { "title" => "Ground Shipping" } }),
    )
    _(body["requestedShippingService"]).must_equal "Ground Shipping"
    _(body.key?("carrierCode")).must_equal false
    _(body.key?("serviceCode")).must_equal false
  end

  test "a 4xx rejection records FAILED with the real reason and does not raise" do
    result = nil
    HTTParty.stub(:post, ->(*_a, **_k) { FakeResponse.new({ "Message" => "Invalid serviceCode" }, 400) }) do
      FluidApi::V2::OrdersService.stub(:new, FakeOrdersService.new) do
        result = Shipstation::CreateOrder.new(order_params(status: "awaiting_shipment")).call
      end
    end
    _(result.success?).must_equal false
    order = @company.shipstation_orders.find_by(fluid_order_id: 555)
    _(order.status).must_equal "FAILED"
    _(order.last_error).must_equal "ShipStation 400: Invalid serviceCode"
    _(order.retry_count).must_equal 1
  end

  test "a 5xx failure raises so the job retries (transient, unlike a 4xx)" do
    error = assert_raises(RuntimeError) do
      HTTParty.stub(:post, ->(*_a, **_k) { FakeResponse.new({ "Message" => "server error" }, 500) }) do
        FluidApi::V2::OrdersService.stub(:new, FakeOrdersService.new) do
          Shipstation::CreateOrder.new(order_params(status: "awaiting_shipment")).call
        end
      end
    end
    _(error.message).must_include "ShipStation 500: server error"
  end

  # -- unfulfillable / cancellation -----------------------------------------

  test "cancels a submitted order in ShipStation when Fluid marks it unfulfillable" do
    @company.shipstation_orders.create!(
      fluid_order_id: 555, fluid_order_number: "ORD-555", status: "SUBMITTED", shipstation_order_id: "999",
    )
    Shipstation::CancelOrder.stub(:new, ->(*) { FakeCancel.new(:cancelled) }) do
      Shipstation::CreateOrder.new(order_params(status: "cancelled")).call
    end
    _(@company.shipstation_orders.find_by(fluid_order_id: 555).status).must_equal "CANCELLED"
  end

  test "does NOT cancel a submitted order that already has a label" do
    order = @company.shipstation_orders.create!(
      fluid_order_id: 555, fluid_order_number: "ORD-555", status: "SUBMITTED", shipstation_order_id: "999",
    )
    Shipstation::CancelOrder.stub(:new, ->(*) { FakeCancel.new(:skipped_has_label) }) do
      Shipstation::CreateOrder.new(order_params(status: "cancelled")).call
    end
    order.reload
    _(order.status).must_equal "SUBMITTED"
    _(order.last_error).must_include "already has a label"
  end

  test "marks a pre-submit order CANCELLED without calling ShipStation" do
    @company.shipstation_orders.create!(
      fluid_order_id: 555, fluid_order_number: "ORD-555", status: "HELD",
    )
    Shipstation::CancelOrder.stub(:new, ->(*) { raise "should not call ShipStation" }) do
      Shipstation::CreateOrder.new(order_params(status: "cancelled")).call
    end
    _(@company.shipstation_orders.find_by(fluid_order_id: 555).status).must_equal "CANCELLED"
  end

  test "never recalls an already-SHIPPED order" do
    order = @company.shipstation_orders.create!(
      fluid_order_id: 555, fluid_order_number: "ORD-555", status: "SHIPPED", shipstation_order_id: "999",
    )
    Shipstation::CancelOrder.stub(:new, ->(*) { raise "should not call ShipStation" }) do
      Shipstation::CreateOrder.new(order_params(status: "cancelled")).call
    end
    _(order.reload.status).must_equal "SHIPPED"
  end
end

File: src/lib/shipstation/create-order.test.ts
/**
 * createShipstationOrder — the decision table.
 *
 * The case that matters most, and the reason several of these tests exist at
 * all: **this droplet must never modify an order that already has a label.**
 * ShipStation moves a labeled order to "shipped", so both paths that could
 * touch a live ShipStation order ask about that status first and back off.
 */

import { describe, it, expect, vi, beforeEach } from "vitest";

const mockPrisma = vi.hoisted(() => ({
  company: { findFirst: vi.fn() },
  integrationSetting: { findUnique: vi.fn() },
  shipstationOrder: {
    findUnique: vi.fn(),
    findUniqueOrThrow: vi.fn(),
    findFirst: vi.fn(),
    create: vi.fn(),
    update: vi.fn(),
  },
  shippingMethodMapping: { findFirst: vi.fn() },
  seenShippingMethod: { findFirst: vi.fn(), create: vi.fn(), update: vi.fn() },
  $transaction: vi.fn(),
  $queryRaw: vi.fn(),
}));

const isShipped = vi.hoisted(() => vi.fn(async () => false));
const cancelOrder = vi.hoisted(() =>
  vi.fn<() => Promise<"cancelled" | "skipped_has_label" | "already_cancelled" | "not_found">>(
    async () => "cancelled",
  ),
);
const updateExternalId = vi.hoisted(() => vi.fn(async () => {}));

vi.mock("@/lib/db", () => ({ prisma: mockPrisma, default: mockPrisma }));
vi.mock("./order-status", () => ({ isShipstationOrderShipped: isShipped }));
vi.mock("./cancel-order", () => ({ cancelShipstationOrder: cancelOrder }));
vi.mock("@/lib/fluid-api", () => ({ updateOrderExternalId: updateExternalId }));
vi.mock("./client", async (importOriginal) => {
  const actual = await importOriginal<typeof import("./client")>();
  return { ...actual, credentialsFor: async () => ({ apiKey: "k", apiSecret: "s" }) };
});

const { createShipstationOrder } = await import("./create-order");

const COMPANY = {
  id: 1n,
  name: "Acme",
  fluidCompanyId: 42n,
  authenticationToken: "cat_acme",
};

const SETTING = {
  id: 7n,
  companyId: 1n,
  holdForBatch: false,
  batchWindowMinutes: null,
  apiVersion: "v1",
  storeId: null,
  settings: null,
};

function localOrder(overrides: Record<string, unknown> = {}) {
  return {
    id: 10n,
    companyId: 1n,
    fluidOrderId: 555n,
    fluidOrderNumber: "SO-1",
    shipstationOrderId: null,
    status: "PENDING",
    retryCount: 0,
    holdUntil: null,
    trackingNumbers: [],
    requestPayload: {},
    ...overrides,
  };
}

const ORDER = {
  id: 555,
  order_number: "SO-1",
  status: "awaiting_shipment",
  ship_to: { name: "Jo", address1: "1 Main St", city: "Austin", state: "TX" },
  items: [{ id: 1, sku: "SKU", title: "Widget", quantity: 2, price: "9.99" }],
  metadata: { shipping: { title: "Ground Shipping" } },
};

const input = (overrides: Record<string, unknown> = {}) => ({
  order: { ...ORDER, ...overrides },
  company_id: 42,
});

beforeEach(() => {
  vi.clearAllMocks();
  mockPrisma.company.findFirst.mockResolvedValue(COMPANY);
  mockPrisma.integrationSetting.findUnique.mockResolvedValue(SETTING);
  mockPrisma.shippingMethodMapping.findFirst.mockResolvedValue(null);
  mockPrisma.seenShippingMethod.findFirst.mockResolvedValue(null);
  mockPrisma.seenShippingMethod.create.mockResolvedValue({});
  mockPrisma.shipstationOrder.update.mockImplementation(async ({ data }) => ({
    ...localOrder(),
    ...data,
  }));
  // The row lock is a raw statement inside an interactive transaction; here the
  // transaction just runs its callback.
  mockPrisma.$transaction.mockImplementation(async (fn: (tx: unknown) => unknown) =>
    fn(mockPrisma),
  );
  mockPrisma.$queryRaw.mockResolvedValue([]);
  isShipped.mockResolvedValue(false);
  vi.stubGlobal("fetch", vi.fn());
});

/** A createorder response. */
function shipstationAccepts(orderId = 987) {
  vi.mocked(fetch).mockResolvedValue(
    new Response(JSON.stringify({ orderId }), { status: 200 }),
  );
}

describe("never modifying a labeled order", () => {
  it("does not re-push a changed order once ShipStation has labeled it", async () => {
    // Already SUBMITTED, with a payload that differs from the incoming one.
    const existing = localOrder({
      status: "SUBMITTED",
      shipstationOrderId: "987",
      requestPayload: { ...ORDER, ship_to: { name: "Someone Else" } },
    });
    mockPrisma.shipstationOrder.findUnique.mockResolvedValue(existing);
    mockPrisma.shipstationOrder.findUniqueOrThrow.mockResolvedValue(existing);
    isShipped.mockResolvedValue(true);

    const result = await createShipstationOrder(input());

    expect(result.data).toEqual({ skipped: "labeled" });
    expect(fetch).not.toHaveBeenCalled();
  });

  it("does re-push a changed order that is still open", async () => {
    const existing = localOrder({
      status: "SUBMITTED",
      shipstationOrderId: "987",
      requestPayload: { ...ORDER, ship_to: { name: "Someone Else" } },
    });
    mockPrisma.shipstationOrder.findUnique.mockResolvedValue(existing);
    mockPrisma.shipstationOrder.findUniqueOrThrow.mockResolvedValue(existing);
    isShipped.mockResolvedValue(false);
    shipstationAccepts();

    const result = await createShipstationOrder(input());

    expect(result.success).toBe(true);
    expect(fetch).toHaveBeenCalledOnce();
  });

  it("leaves an unchanged submitted order alone without even asking", async () => {
    const existing = localOrder({
      status: "SUBMITTED",
      shipstationOrderId: "987",
      requestPayload: ORDER,
    });
    mockPrisma.shipstationOrder.findUnique.mockResolvedValue(existing);
    mockPrisma.shipstationOrder.findUniqueOrThrow.mockResolvedValue(existing);

    const result = await createShipstationOrder(input());

    expect(result.data).toEqual({ skipped: "no change" });
    expect(isShipped).not.toHaveBeenCalled();
  });

  it("never touches an order this droplet already knows shipped", async () => {
    const existing = localOrder({ status: "SHIPPED", shipstationOrderId: "987" });
    mockPrisma.shipstationOrder.findUnique.mockResolvedValue(existing);
    mockPrisma.shipstationOrder.findUniqueOrThrow.mockResolvedValue(existing);

    const result = await createShipstationOrder(input());

    expect(result.data).toEqual({ skipped: "already SHIPPED" });
    expect(fetch).not.toHaveBeenCalled();
  });

  it("does not cancel a labeled order when Fluid says it is unfulfillable", async () => {
    mockPrisma.shipstationOrder.findFirst.mockResolvedValue(
      localOrder({ status: "SUBMITTED", shipstationOrderId: "987" }),
    );
    cancelOrder.mockResolvedValue("skipped_has_label");

    const result = await createShipstationOrder(input({ status: "cancelled" }));

    expect(result.data).toEqual({ skipped: "status=cancelled" });
    // The local record keeps its status and records why it was left alone.
    const update = mockPrisma.shipstationOrder.update.mock.calls.at(-1)?.[0];
    expect(update.data.status).toBeUndefined();
    expect(update.data.lastError).toContain("already has a label");
  });

  it("does cancel an unfulfillable order that has no label yet", async () => {
    mockPrisma.shipstationOrder.findFirst.mockResolvedValue(
      localOrder({ status: "SUBMITTED", shipstationOrderId: "987" }),
    );
    cancelOrder.mockResolvedValue("cancelled");

    await createShipstationOrder(input({ status: "refunded" }));

    const update = mockPrisma.shipstationOrder.update.mock.calls.at(-1)?.[0];
    expect(update.data.status).toBe("CANCELLED");
  });
});

describe("status gating", () => {
  beforeEach(() => {
    const fresh = localOrder();
    mockPrisma.shipstationOrder.findUnique.mockResolvedValue(null);
    mockPrisma.shipstationOrder.create.mockResolvedValue(fresh);
    mockPrisma.shipstationOrder.findUniqueOrThrow.mockResolvedValue(fresh);
  });

  it("holds an unpaid order instead of sending it", async () => {
    const result = await createShipstationOrder(input({ status: "awaiting_payment" }));

    expect(result.data).toEqual({ held: true });
    expect(fetch).not.toHaveBeenCalled();
  });

  it("batch-holds when the company has batching on", async () => {
    mockPrisma.integrationSetting.findUnique.mockResolvedValue({
      ...SETTING,
      holdForBatch: true,
      batchWindowMinutes: 15,
    });

    const result = await createShipstationOrder(input());

    expect(result.data).toEqual({ held_for_batch: true });
    expect(fetch).not.toHaveBeenCalled();
  });

  it("bypasses the batching hold when told to", async () => {
    mockPrisma.integrationSetting.findUnique.mockResolvedValue({
      ...SETTING,
      holdForBatch: true,
    });
    shipstationAccepts();

    const result = await createShipstationOrder(input(), { respectHold: false });

    expect(result.success).toBe(true);
    expect(fetch).toHaveBeenCalledOnce();
  });

  it("records a 4xx as FAILED without throwing, because a retry cannot fix it", async () => {
    vi.mocked(fetch).mockResolvedValue(
      new Response(JSON.stringify({ Message: "Invalid serviceCode" }), { status: 400 }),
    );

    const result = await createShipstationOrder(input());

    expect(result.success).toBe(false);
    expect(result.error).toBe("ShipStation 400: Invalid serviceCode");
    const update = mockPrisma.shipstationOrder.update.mock.calls.at(-1)?.[0];
    expect(update.data.status).toBe("FAILED");
  });

  it("throws on a 5xx, so the webhook answers 500 and Fluid retries", async () => {
    vi.mocked(fetch).mockResolvedValue(new Response("boom", { status: 503 }));

    await expect(createShipstationOrder(input())).rejects.toThrow(/ShipStation error/);
  });
});

describe("the ShipStation payload", () => {
  beforeEach(() => {
    const fresh = localOrder();
    mockPrisma.shipstationOrder.findUnique.mockResolvedValue(null);
    mockPrisma.shipstationOrder.create.mockResolvedValue(fresh);
    mockPrisma.shipstationOrder.findUniqueOrThrow.mockResolvedValue(fresh);
    shipstationAccepts();
  });

  const sentBody = () =>
    JSON.parse(String(vi.mocked(fetch).mock.calls[0][1]?.body));

  it("sends the shipping title even with no mapping configured", async () => {
    await createShipstationOrder(input());

    const body = sentBody();
    expect(body.requestedShippingService).toBe("Ground Shipping");
    expect(body.carrierCode).toBeUndefined();
  });

  it("sends carrier and service only as a complete pair", async () => {
    mockPrisma.shippingMethodMapping.findFirst.mockResolvedValue({
      carrierCode: "stamps_com",
      serviceCode: null,
      packageCode: null,
    });

    await createShipstationOrder(input());

    // ShipStation rejects a carrier with no service ("Invalid serviceCode"), so
    // the title alone goes rather than a 400.
    expect(sentBody().carrierCode).toBeUndefined();
  });

  it("assigns the configured store", async () => {
    mockPrisma.integrationSetting.findUnique.mockResolvedValue({
      ...SETTING,
      storeId: "12345",
    });

    await createShipstationOrder(input());

    expect(sentBody().advancedOptions).toEqual({ storeId: 12345 });
  });

  it("puts the ordered variant on the line item so the packing slip shows it", async () => {
    await createShipstationOrder(
      input({
        items: [
          {
            id: 1,
            title: "Widget",
            quantity: 1,
            ordered_variant: [{ option_type: "Size", value: "Large" }],
          },
        ],
      }),
    );

    const item = sentBody().items[0];
    expect(item.options).toEqual([{ name: "Size", value: "Large" }]);
    expect(item.name).toBe("Widget — Large");
  });

  it("ignores Fluid's Default Variant placeholder", async () => {
    await createShipstationOrder(
      input({
        items: [
          { id: 1, title: "Widget", quantity: 1, variant: { title: "Default Variant" } },
        ],
      }),
    );

    const item = sentBody().items[0];
    expect(item.options).toBeUndefined();
    expect(item.name).toBe("Widget");
  });

  it("converts a kg weight to grams and omits an absent one", async () => {
    await createShipstationOrder(
      input({
        items: [
          { id: 1, title: "A", quantity: 1, weight: 1.5, unit_of_weight: "kg" },
          { id: 2, title: "B", quantity: 1 },
        ],
      }),
    );

    const items = sentBody().items;
    expect(items[0].weight).toEqual({ value: 1500, units: "grams" });
    expect(items[1].weight).toBeUndefined();
  });
});
